### PR TITLE
[None][fix] Fix encoder-decoder beam search corruption (cross-KV sharing, mixed-beam exception, gather_generation_logits)

### PR DESCRIPTION
**Pull request by achartier**
@coderabbitai summary

## Description

Fixes three bugs in the TRT backend (`ModelRunnerCpp`) that caused corrupted output when concurrent requests with different beam widths (e.g. beam=1 and beam=5) were submitted to the same executor — specifically for encoder-decoder models (Whisper) with beam search.

**Bug 1 — Cross-KV block sharing (Fix 1)**

`KVCacheManager::copyBlockOffsets` assigned separate physical blocks to each beam of an encoder-decoder cross-attention sequence. The TRT encoder runs with `contextBeamWidth=1` so only beam-0's blocks were ever populated with encoder features. Beams 1..N-1 attended to uninitialised GPU memory, producing degenerate repetitive output ("happ happ happ").

Fix: when `isCrossKv()` is true, always use beam-0's source block IDs and count for all beams. The encoder output is identical for all beams of a request.

**Bug 2 — `mInflightReqIds` leak on mixed-beam exception (Fix 2)**

When concurrent requests with different beam widths triggered the `verifyRequests()` exception in `TrtGptModelInflightBatching::forwardAsync`, the exception handler freed the KV-cache but did not remove request IDs from `mInflightReqIds`. The subsequent `changeBeamWidth()` call checks `TLLM_CHECK(mInflightReqIds.empty())` and aborted.

Fix: erase affected IDs from `mInflightReqIds` and call `changeBeamWidth(mOperatingBeamWidth)` in the exception catch block so the next batch starts from a clean state.

**Bug 3 — `gather_generation_logits` corruption with concurrent beam widths (Fix 3)**

`copyGenerationLogits()` used `mergeLogitsFragmentsKernel` — a CUDA kernel that reads fragment GPU addresses via a shared `fragmentPointerDevice` pointer array — to gather kCACHE_LENGTH logit steps into a scratch buffer before copying to host. When `gather_generation_logits=True` was combined with concurrent mixed beam-width requests (causing `changeBeamWidth()` to be called between batches), this kernel intermittently corrupted token generation, producing ~26% degenerate output.

Fix: bypass the kernel entirely. Copy each (beam, step) logit slice directly from its fragment GPU tensor to the corresponding host location via `bufferManager.copy()`. This is functionally equivalent and eliminates the pointer-indirection pattern that caused the corruption.

The dead kernel (`mergeLogitsFragmentsKernel`, `invokeMergeLogitsFragments`, `mergeLogitsFragments`) and its now-unused scratch buffers (`transposedLogits`, `fragmentPointerDevice`, `fragmentPointerHost`) are removed in commit 4.

## Test Coverage

**C++ unit tests** (`cpp/tests/unit_tests/batch_manager/encDecBeamSearchTest.cpp`):
- `CrossKvBeamSharingTest/CopyBlockOffsetsAllBeamsShareBeam0Blocks` — verifies Fix 1 invariant: all beam slots in the cross-KV offset table equal beam-0.
- `CopyGenerationLogitsTest/DirectCopyPlacesEachBeamStepAtCorrectHostOffset` — verifies Fix 3: each (beam, step) logit lands at the correct host offset. Making `copyGenerationLogits()` a no-op produces 56 assertion failures.

**Validation on Whisper large-v3 with concurrent mixed-beam execution:**
| Configuration | Before fixes | After fixes |
|---|---|---|
| beam=5 only, no mixing | — | PASS 10/10 |
| mix + no `gather_generation_logits` | crash/corrupt | **PASS 20/20** |
| mix + `gather_generation_logits=True` | crash/corrupt | **PASS 50/50** |

## PR Checklist

- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: cpp/tensorrt_llm/batch_manager/utils/inflightBatchingUtils.cpp [hunk 1]
```diff
@@ -102,34 +101,31 @@ void copyGenerationLogits(RuntimeBuffers::GenerationLogitsCache& generationLogit
...
+        for (SizeType32 stepIdx = 0; stepIdx < static_cast<SizeType32>(beamFragmentSize); ++stepIdx)
+        {
+            // frag shape: [1, beamWidth, vocabSizePadded]. Beam b starts at offset b*vocab.
+            auto fragBeamSlice = ITensor::slice(fragments.at(stepIdx), {0, beam}, 1);
+            // host shape: [beamWidth, mMaxNewTokens, vocabSizePadded]. Target: [beam, hostOffset+stepIdx, :].
+            auto hostStepSlice = ITensor::slice(llmReq.getGenerationLogitsHost(), {beam, hostOffset + stepIdx}, 1);
+            bufferManager.copy(*fragBeamSlice, *hostStepSlice);
+        }
     }
+    llmReq.clearGenerationLogitsFragments();
```
[the remark this answers is not in the document]
Comment (achartier) [reply cpp/tensorrt_llm/batch_manager/utils/inflightBatchingUtils.cpp]: Should be fixed in the last version, please check.

[the remark this answers is not in the document]
Comment (achartier) [reply cpp/tensorrt_llm/batch_manager/utils/inflightBatchingUtils.cpp]: We've reconsidered and removed the synchronize() call (see latest commit) for the following reasons:

1. **No actual use-after-free**: `clearGenerationLogitsFragments()` releases the fragment tensor *objects* (shared_ptr reference counts), but the underlying GPU buffer is owned by `generationLogitsCache.logits` inside `RuntimeBuffers`, which outlives the fragment objects until the next `changeBeamWidth()` call. The async copies read from this buffer — they are not affected by the shared_ptr teardown.

2. **GPU ordering is already guaranteed**: `GptDecoderBatched::forwardAsync` records a CUDA event on the runtime stream after this function returns, and the decoder stream waits on that event before reading any shared GPU state. So even without a CPU-side synchronize, all copies are guaranteed to complete before anything downstream reads or frees the source data.

3. **Performance**: `bufferManager.getStream().synchronize()` would block the CPU thread every 8 generation steps, turning an async pipeline stage into a synchronous one. Given the latency impact we opted to remove it and document the reasoning in a comment instead.

Does this analysis hold? Happy to revisit if there is a scenario we've missed.

**Comment by achartier** [comment]
Closing - better fix implemented in https://github.com/NVIDIA/TensorRT-LLM/pull/15461
